src/kafka/producer/sender.cc

```diff
@@ -0,0 +1,261 @@
+                topic_data._partitions->push_back(partition_data);
+            }
+            req._topics->push_back(topic_data);
+        }
+
+        _responses.emplace_back(_connection_manager->send(req, broker.first, broker.second)
+            .then([broker](auto response) {
+                return std::make_pair(broker, response);
+        }));
+    }
+}
+
+void sender::set_error_code_for_broker(const sender::connection_id& broker, const error::kafka_error_code& error_code) {
+    for (auto& [topic, messages_by_partition] : _messages_split_by_broker_topic_partition[broker]) {
+        for (auto& [partition, messages] : messages_by_partition) {
+            for (auto& message : messages) {
+                message->_error_code = error_code;
+            }
+        }
+    }
+}
+
+void sender::set_error_code_for_topic_partition(const std::string& topic, int32_t partition_index,
+        const error::kafka_error_code& error_code) {
+    for (auto& message : _messages_split_by_topic_partition[{topic, partition_index}]) {
+        message->_error_code = error_code;
+    }
+}
+
+void sender::set_success_for_topic_partition(const std::string& topic, int32_t partition_index) {
+    for (auto& message : _messages_split_by_topic_partition[{topic, partition_index}]) {
+        message->_error_code = error::kafka_error_code::NONE;
+        message->_promise.set_value();
+    }
+}
+
+void sender::move_messages(std::vector<sender_message>& messages) {
+    _messages.insert(_messages.end(), std::make_move_iterator(messages.begin()),
+              std::make_move_iterator(messages.end()));
+    messages.clear();
+}
+
+size_t sender::messages_size() const {
+    return _messages.size();
+}
+
+bool sender::messages_empty() const {
+    return _messages.empty();
+}
+
+void sender::send_requests() {
+    split_messages();
+    queue_requests();
+}
+
+future<> sender::receive_responses() {
+    return when_all(_responses.begin(), _responses.end()).then([this](auto responses){
+        set_error_codes_for_responses(responses);
+        filter_messages();
+        return process_messages_errors();
+    });
+}
+
+future<> sender::process_messages_errors() {
+    auto should_refresh_metadata = false;
+    for (auto& message : _messages) {
+        if ((*message._error_code)._is_invalid_metadata) {
+            should_refresh_metadata = true;
+            break;
+        }
+    }
+    if (should_refresh_metadata) {
+        return _metadata_manager->refresh_metadata().discard_result();
+    } else {
+        return make_ready_future<>();
+    }
+}
+
+void sender::filter_messages() {
+    _messages.erase(std::remove_if(_messages.begin(), _messages.end(), [](auto& message) {
+        if (message._error_code == error::kafka_error_code::NONE) {
+            return true;
+        }
+        if (!(*message._error_code)._is_retriable) {
+            message._promise.set_exception(send_exception((*message._error_code)._error_message));
+            return true;
+        }
```

Comment:
Just to make sure, if a message runs out of retries `_is_retriable` will be set to `false`, correct?

Reply:
No, `with_retry` runs inner lambda N times (it controls running out of retries). It doesn't change `_is_retriable`. 

`_is_retriable` describes whether retrying makes "sense" - in case of `NETWORK_EXCEPTION` you can reconnect and try again, but in case of `MESSAGE_TOO_LARGE` retrying won't change the outcome as the message still will be too large. `_is_retriable` is constant for each error code.

Reply:
So you filter it out if it's retriable, which makes sense for a message we can attempt to send again, but wouldn't we want to have it there, not removed, if we ran out of retries and still get an error?

Reply:
I filter out non-retriable errors.

```
 if (!(*message._error_code)._is_retriable) {
```

In case of running out of retries, `sender.close()` is ran after, which sets exceptions for out-of-retries messages.

Reply:
My bad, I didn't notice the exclamation mark